driver-core/src/main/com/mongodb/connection/LoggingCommandEventSender.java

```diff
@@ -0,0 +1,172 @@
+import com.mongodb.diagnostics.logging.Logger;
+import com.mongodb.event.CommandListener;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonReader;
+import org.bson.codecs.RawBsonDocumentCodec;
+import org.bson.json.JsonMode;
+import org.bson.json.JsonWriter;
+import org.bson.json.JsonWriterSettings;
+
+import java.io.StringWriter;
+import java.util.Set;
+
+import static com.mongodb.connection.ProtocolHelper.sendCommandFailedEvent;
+import static com.mongodb.connection.ProtocolHelper.sendCommandStartedEvent;
+import static com.mongodb.connection.ProtocolHelper.sendCommandSucceededEvent;
+import static java.lang.String.format;
+
+class LoggingCommandEventSender implements CommandEventSender {
+    private static final int MAX_COMMAND_DOCUMENT_LENGTH_TO_LOG = 1000;
+
+    private final Set<String> securitySensitiveCommands;
+    private final ConnectionDescription description;
+    private final CommandListener commandListener;
+    private final Logger logger;
+    private final long startTimeNanos;
+    private final CommandMessage message;
+    private final String commandName;
+    private volatile BsonDocument commandDocument;
+
+    LoggingCommandEventSender(final Set<String> securitySensitiveCommands, final ConnectionDescription description,
+                              final CommandListener commandListener, final CommandMessage message,
+                              final ByteBufferBsonOutput bsonOutput, final Logger logger) {
+        this.securitySensitiveCommands = securitySensitiveCommands;
+        this.description = description;
+        this.commandListener = commandListener;
+        this.logger = logger;
+        this.startTimeNanos = System.nanoTime();
+        this.message = message;
+        this.commandDocument = message.getCommandDocument(bsonOutput);
+        this.commandName = commandDocument.getFirstKey();
+    }
+
+    @Override
+    public void sendStartedEvent() {
+        if (loggingRequired()) {
+            logger.debug(
+                    format("Sending command '%s ...' with request id %d to database %s on connection [%s] to server %s",
+                            getTruncatedJsonCommand(), message.getId(),
+                            message.getNamespace().getDatabaseName(), description.getConnectionId(), description.getServerAddress()));
+        }
+
+        if (eventRequired()) {
+            BsonDocument commandDocumentForEvent = (securitySensitiveCommands.contains(commandName))
+                    ? new BsonDocument() : commandDocument;
+
+            sendCommandStartedEvent(message, message.getNamespace().getDatabaseName(),
+                    commandName, commandDocumentForEvent, description, commandListener);
+        }
+        // the buffer underlying the command document may be released after the started event, so set to null to ensure it's not used
+        // when sending the failed or succeeded event
+        commandDocument = null;
```

Comment:
This is a wart that I'm not sure how to get rid of cleanly

Reply:
Ack